### PR TITLE
[AutoParallel] Add local view reshape and nd_mesh_alltoall reshard for moe

### PR DESCRIPTION
**Pull request by pkuzyc**
### PR Category
Auto Parallel

### PR Types
New features

### Description
Pcard-67164

All operations in auto parallel are identical to the serial case, i.e. all operations are global view. In MoE, the global view reshape brings additional communication. 

This pr add local view reshape and nd_mesh_alltoall reshard to make the MoE training the same as original distributed training, to achieve higher performance.

Additional: add deepcopy function for placements.

### CODE REVIEW

**Review by winter-wang — DISMISSED** [review]

File: python/paddle/distributed/auto_parallel/moe_utils.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+def _cal_local_shape(global_shape, mesh, placements):
+    local_shape = list(global_shape)
+    for idx, placement in enumerate(placements):
+        if placement.is_shard():
+            shard_dim = placement.get_dim()
+            local_shape[shard_dim] = local_shape[shard_dim] // mesh.shape[idx]
+    return local_shape
+
+
+def infer_pos_shape(src_shape, tgt_shape):
```
Comment (From00) [inline python/paddle/distributed/auto_parallel/moe_utils.py]: pos是什么的缩写

File: python/paddle/distributed/auto_parallel/moe_utils.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+        ret_shape = tgt_shape.copy()
+
+    neg_one_idx = np.where(ret_shape == -1)[0]
+    if neg_one_idx.size > 0:
+        assert (
+            neg_one_idx.size <= 1
+        ), "At most one -1 is allowed in target shape."
+
+        nelem = np.prod(src_shape)
+        ret_shape[neg_one_idx[0]] = 1
```
Comment (From00) [inline python/paddle/distributed/auto_parallel/moe_utils.py]: 这行代码是冗余的

File: python/paddle/distributed/auto_parallel/moe_utils.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+
+
+def _reshard_mesh_shape(
+    dist_tensor: Tensor, mesh: ProcessMesh, placements: list[Placement]
+):
+    src_mesh = dist_tensor.process_mesh
+    if src_mesh == mesh or src_mesh.process_ids != mesh.process_ids:
+        return False
+
+    # only the mesh shapes are different,
```
Comment (From00) [inline python/paddle/distributed/auto_parallel/moe_utils.py:222]: 这种情况是否可以在dist_reshape中也加一个拦截报错，防止用户直接调用dist_reshape产生非预期行为。

File: python/paddle/distributed/auto_parallel/api.py [hunk 4]
```diff
@@ -772,6 +777,17 @@ def reshard(
         if len(partial_dims) > 0:
             dist_attr._set_partial_dims(partial_dims)
 
+        alltoall_dim = _specific_alltoall_dim(dist_tensor, mesh, placements)
+        if alltoall_dim is not None:
+            # return _nd_mesh_alltoall_reshard(dist_tensor, mesh, placements, alltoall_dim)
```
Comment (From00) [inline python/paddle/distributed/auto_parallel/api.py]: 这行代码是否可以删除

File: paddle/phi/core/distributed/auto_parallel/dist_tensor.cc [hunk 5]
```diff
@@ -49,16 +49,16 @@ TensorDistAttr ToTensorDistAttr(const ProcessMesh& process_mesh,
...
-          dim_map[shard_dim],
-          -1,
-          common::errors::InvalidArgument(
-              "Tensor dim %lld is already sharded on mesh dim %lld,"
-              " DistTensor operator implementation does not support things "
-              "like hybrid"
-              " sharding strategies yet (i.e. [Shard(0), Shard(0)])",
-              shard_dim,
-              dim_map[shard_dim]));
+      // PADDLE_ENFORCE_EQ(
```
Comment (From00) [inline paddle/phi/core/distributed/auto_parallel/dist_tensor.cc:59]: 这行检查改成WARNING会不会好一些

Comment (pkuzyc) [reply paddle/phi/core/distributed/auto_parallel/dist_tensor.cc:59]: done

File: python/paddle/distributed/auto_parallel/moe_utils.py (hunk 3 above)
Comment (pkuzyc) [reply python/paddle/distributed/auto_parallel/moe_utils.py:222]: done

File: python/paddle/distributed/auto_parallel/moe_utils.py (hunk 1 above)
Comment (pkuzyc) [reply python/paddle/distributed/auto_parallel/moe_utils.py]: positive，改成 positive了

File: python/paddle/distributed/auto_parallel/moe_utils.py (hunk 2 above)
Comment (pkuzyc) [reply python/paddle/distributed/auto_parallel/moe_utils.py]: 这个方法用来推断输入shape里面的 -1 对应的值是多少。这行把 -1 设成 1 是为了下面 np.prod(ret_shape) 算出来是个正数。

File: python/paddle/distributed/auto_parallel/api.py (hunk 4 above)
Comment (pkuzyc) [reply python/paddle/distributed/auto_parallel/api.py]: done

**Review by XieYunshen — APPROVED** [review]

**Review by winter-wang — APPROVED** [review]

**Review by From00 — APPROVED** [review]
LGTM

**Pull request merged**